src/object.rs

```diff
@@ -714,4 +717,56 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
             }
         }
     }
+
+    fn downgrade(&self) -> WeakRef<T> {
+        unsafe {
+            let w = WeakRef(Box::new(mem::uninitialized()), PhantomData);
+            gobject_ffi::g_weak_ref_init(mut_override(&*w.0), self.to_glib_none().0);
+            w
+        }
+
+    }
+}
+
+pub struct WeakRef<T: IsA<Object> + ?Sized>(Box<gobject_ffi::GWeakRef>, PhantomData<*const T>);
+
+impl<T: IsA<Object> + Cast + ?Sized> WeakRef<T> {
+    pub fn upgrade(&self) -> Option<T> {
+        unsafe {
+            let ptr = gobject_ffi::g_weak_ref_get(mut_override(&*self.0));
+            if ptr.is_null() {
+                None
+            } else {
+                let obj: Object = from_glib_full(ptr);
+                Some(Cast::downcast::<T>(obj).expect("Wrongly typed GWeakRef"))
+            }
+        }
+    }
 }
+
+impl<T: IsA<Object> + ?Sized> Drop for WeakRef<T> {
+    fn drop(&mut self) {
+        unsafe {
+            gobject_ffi::g_weak_ref_clear(mut_override(&*self.0));
+        }
+    }
+}
+
+impl<T: IsA<Object> + ?Sized> Clone for WeakRef<T> {
+    fn clone(&self) -> Self {
+        unsafe {
+            let c = WeakRef(Box::new(mem::uninitialized()), PhantomData);
+
+            let o = gobject_ffi::g_weak_ref_get(mut_override(&*self.0));
+            gobject_ffi::g_weak_ref_init(mut_override(&*self.0), o);
```

Comment:
I had no Clone in my code before and only added that for consistency with Arc now. So that part was untested. But easy to fix, will do so later